### PR TITLE
Use separate prevalence estimates for each sample

### LINKED ISSUE #46: Estimate viral prevalence separately for each MGS sample
For some viruses, e.g. SARS-CoV2, we'll have multiple estimates for different times and locations. We'll want to:

1. Allow the prevalence estimate (and standard error) to be a vector, with one estimate per sample.
2. Figure out how to get one MGS estimate for each sample. This will probably be some simple matching at first, but ultimately we may want to model explicitly.

### PR DESCRIPTION
**Pull request by dp-rice**
Resolves #46. Changes the statistical model to have a separate true prevalence parameter for each sample. At the moment they are independent, but future updates should allow for correlations. The provided estimated mean log prevalence can be either a scalar or a vector. If a scalar, each sample has the same prior mean. If a vector, each sample has the corresponding prior mean.

Fitting to the Rothman SARS-CoV-2 data with per-sample prevalence estimates:
```
> ./fit_covid.py 2> /dev/null

----------------------------------------------
sars_cov_2 relative abundance at 1% prevalence
----------------------------------------------
Naive estimate:
3.4e-07
Posterior arithmetic mean:
1.2e-07
Posterior geometric mean:
1.2e-07
Posterior quantiles:
     5%        25%        50%        75%        95%
8.0e-08    9.9e-08    1.1e-07    1.3e-07    1.7e-07
```

Fitting to all viruses with only a single mean prevalence estimate. Note that these results are different to #44 because the model has changed:
```
>  ./fit_rothman.py 2> /dev/null

----------------------------------------------
sars_cov_2 relative abundance at 1% prevalence
----------------------------------------------
Naive estimate:
2.9e-07
Posterior arithmetic mean:
1.2e-07
Posterior geometric mean:
1.1e-07
Posterior quantiles:
     5%        25%        50%        75%        95%
6.6e-08    8.9e-08    1.1e-07    1.4e-07    2.1e-07


---------------------------------------
hiv relative abundance at 1% prevalence
---------------------------------------
Naive estimate:
7.8e-09
Posterior arithmetic mean:
6.9e-09
Posterior geometric mean:
4.1e-09
Posterior quantiles:
     5%        25%        50%        75%        95%
5.4e-10    2.2e-09    4.9e-09    9.0e-09    2.0e-08


-----------------------------------------
hsv_2 relative abundance at 1% prevalence
-----------------------------------------
Naive estimate:
0.0e+00
Posterior arithmetic mean:
2.4e-11
Posterior geometric mean:
3.7e-12
Posterior quantiles:
     5%        25%        50%        75%        95%
1.1e-14    9.6e-13    7.2e-12    2.8e-11    1.0e-10


-----------------------------------------
hsv_1 relative abundance at 1% prevalence
-----------------------------------------
Naive estimate:
9.5e-10
Posterior arithmetic mean:
5.8e-10
Posterior geometric mean:
5.6e-10
Posterior quantiles:
     5%        25%        50%        75%        95%
3.6e-10    4.6e-10    5.5e-10    6.7e-10    8.7e-10


---------------------------------------------
norovirus relative abundance at 1% prevalence
---------------------------------------------
Naive estimate:
3.0e-03
Posterior arithmetic mean:
1.8e-03
Posterior geometric mean:
1.7e-03
Posterior quantiles:
     5%        25%        50%        75%        95%
1.2e-03    1.5e-03    1.7e-03    2.0e-03    2.6e-03


---------------------------------------
hav relative abundance at 1% prevalence
---------------------------------------
Naive estimate:
0.0e+00
Posterior arithmetic mean:
3.1e-06
Posterior geometric mean:
1.9e-07
Posterior quantiles:
     5%        25%        50%        75%        95%
1.0e-10    3.2e-08    5.0e-07    2.9e-06    1.5e-05
```

### CODE REVIEW

**Comment by dp-rice** [comment]
One outstanding issue: I don't understand how `numpy` type annotations work. Once I figure that out, I can clean up some of the comments in `stats.py` and `fit_covid.py`.

**Comment by dp-rice** [comment]
Update: just added some minimal numpy annotations. Enough to quiet mypy at least. They could be made better, but it's not worth it until the stats interface is more developed, which is waiting on refining the model.

File: fit_covid.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
+#!/usr/bin/env python3
+import numpy as np
+
+import stats
+from fit_rothman import per100k_to_per100, print_summary
+from mgs import BioProject, MGSData
+from pathogens import pathogens
+
+plant_counties = {
```
Comment (jeffkaufman) [inline fit_covid.py:9]: This is now in the sample metadata, under `county`, though using that can be a later PR

File: fit_covid.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    "HTP": "Los Angeles",
+    "SJ": "Los Angeles",
+    "JWPCP": "Los Angeles",
+    "OC": "Orange",
+    "PL": "San Diego",
+    "SB": "San Diego",
+    "NC": "San Diego",
+}
+
+if __name__ == "__main__":
```
Comment (jeffkaufman) [inline fit_covid.py:19]: Putting lots of stuff under `if __name__ == "__main__":` feels a bit wrong to me, though maybe for a script like this it doesn't actually have downsides?  I would normally do:

```
def start():
  ....

if __name__ == "__main__":
  start()
```

File: fit_covid.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        assert key not in prevalence_by_loc_date
+        prevalence_by_loc_date[key] = estimate.infections_per_100k
+
+    prevalence_per100k = np.zeros(len(samples))
+    for i, (sample, attrs) in enumerate(samples.items()):
+        assert attrs.fine_location is not None
+        try:
+            county = plant_counties[attrs.fine_location]
+        except KeyError:
+            prevalence_per100k[i] = np.nan
```
Comment (jeffkaufman) [inline fit_covid.py:47]: I think this shouldn't happen?

File: fit_covid.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+    prevalence_by_loc_date = {}
+    for estimate in pathogens[pathogen].estimate_prevalences():
+        assert estimate.parsed_start == estimate.parsed_end
+        key = (estimate.county, estimate.parsed_start)
+        assert key not in prevalence_by_loc_date
+        prevalence_by_loc_date[key] = estimate.infections_per_100k
+
+    prevalence_per100k = np.zeros(len(samples))
+    for i, (sample, attrs) in enumerate(samples.items()):
```
Comment (jeffkaufman) [inline fit_covid.py:42]: When you update this to pull the latest data you'll want to either limit to ones with enrichment=viral or calculate separate estimates for viral and panel enrichment.

File: fit_covid.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    "SJ": "Los Angeles",
+    "JWPCP": "Los Angeles",
+    "OC": "Orange",
+    "PL": "San Diego",
+    "SB": "San Diego",
+    "NC": "San Diego",
+}
+
+if __name__ == "__main__":
+    bioproject = BioProject("PRJNA729801")  # Rothman
```
Comment (jeffkaufman) [inline fit_covid.py:20]: If I update our covid prevalence data to include Bay Area counties you should be able to run this for Crits-Christoph too

**Review by jeffkaufman — APPROVED** [review]
Looks reasonable, though again I'm not the person to review the stats in depth (but that's fine and this can go in)

File: fit_covid.py (hunk 2 above)
Comment (dp-rice) [reply fit_covid.py:19]: These fitting scripts are meant to be transient. As we develop the model and apply it to more data, I'm planning on packaging most of this into functions/modules, so I'm not going to worry about this for now.

File: fit_covid.py (hunk 3 above)
Comment (dp-rice) [reply fit_covid.py:47]: I was getting errors for `ESC` and `JW`. This will be addressed when we update to include the county info.

**Pull request merged**